### PR TITLE
MODFIN-110 - Enforce allocatedFromIds and allocatedToIds

### PR DESCRIPTION
**Pull request by siarhei-charniak**
[MODFIN-110](https://issues.folio.org/browse/MODFIN-110) - Enforce allocatedFromIds and allocatedToIds

## Purpose
Allocations and transfers to/from budgets from another budget should be restricted by the fund.allocatedToIds and fund.allocatedFromIds fields in the funds associated with the budgets involved.

## Approach
* Transfers/allocations between budgets are restricted according to requirements
* Unit tests are updated

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/org/folio/rest/impl/TransactionTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  @Test
+  public void testTransferAndAllocationWithEmptyAllocatedIds() throws Exception {
+    logger.info("=== Test transfer and allocation with empty toFund.allocatedFromIds and fromFund.allocatedToIds - Success ===");
+    addMockEntry(FUND.name(), new JsonObject(getMockData("mockdata/funds/CANHIST.json")));
+    addMockEntry(FUND.name(), new JsonObject(getMockData("mockdata/funds/LATHIST.json")));
+
+    // transfer
+    Transaction transaction = createTransaction(TRANSFER)
+      .withFromFundId(CANHIST_ID)
+      .withToFundId(LATHIST_ID);
```
Comment (pautau-s) [inline src/test/java/org/folio/rest/impl/TransactionTest.java:52]: LATHIST fund does not have allocatedTo, that means Transfer  to this fund should not be allowed. And response code must be 422

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 2]
```diff
@@ -90,4 +97,34 @@ private void validateTransactionType(Transaction transaction, Transaction.Transa
...
+      case ALLOCATION:
+      case TRANSFER:
+        if ((Objects.isNull(transaction.getFromFundId()) ^ Objects.isNull(transaction.getToFundId())) &&
+          transaction.getTransactionType().equals(ALLOCATION)) {
+          future.complete(transaction);
+        } else if (Objects.nonNull(transaction.getFromFundId()) && Objects.nonNull(transaction.getToFundId())) {
+          checkAllocatedIds(transaction)
+            .thenApply(isMatch ->
+              Boolean.TRUE.equals(isMatch) ? future.complete(transaction) : future.completeExceptionally(new HttpException(422, ALLOCATION_IDS_MISMATCH)))
+            .exceptionally(throwable -> future.completeExceptionally(new HttpException(500, ALLOCATION_TRANSFER_FAILED)));
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java:112]: ALLOCATION_TRANSFER_FAILED error code is used for Allocation and Transfer, but description says "Failed to create allocation transaction for a budget"

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 3]
```diff
@@ -90,4 +97,34 @@ private void validateTransactionType(Transaction transaction, Transaction.Transa
...
+        future.complete(transaction);
+    }
+    return future;
+  }
+
+  private CompletableFuture<Boolean> checkAllocatedIds(Transaction transaction) {
+    FundsHelper fundsHelper = new FundsHelper(okapiHeaders, ctx, lang);
+    return fundsHelper.getFund(transaction.getFromFundId())
+      .thenCombine(fundsHelper.getFund(transaction.getToFundId()), (from, to) ->
+        (from.getFund().getAllocatedToIds().isEmpty() || from.getFund().getAllocatedToIds().contains(transaction.getToFundId())) &&
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java:127]: According to description if **allocatedTo/allocatedFrom** is empty transfer **from/to** must not be allowed

File: descriptors/ModuleDescriptor-template.json [hunk 4]
```diff
@@ -352,7 +352,12 @@
           "pathPattern": "/finance/allocations",
           "permissionsRequired": ["finance.allocations.item.post"],
           "modulePermissions": [
-            "finance-storage.transactions.item.post"
+            "finance-storage.transactions.item.post",
+            "finance-storage.funds.item.get",
+            "finance-storage.ledgers.item.get",
```
Comment (pautau-s) [inline descriptors/ModuleDescriptor-template.json:357]: I don't see any calls except get fund by id. No need  for following modulesPermissions
```
"finance-storage.fiscal-years.item.get"
 "finance-storage.fiscal-years.collection.get"
"finance-storage.group-fund-fiscal-years.collection.get"
```

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 5]
```diff
@@ -90,4 +97,34 @@ private void validateTransactionType(Transaction transaction, Transaction.Transa
...
+      default:
+        future.complete(transaction);
+    }
+    return future;
+  }
+
+  private CompletableFuture<Boolean> checkAllocatedIds(Transaction transaction) {
+    FundsHelper fundsHelper = new FundsHelper(okapiHeaders, ctx, lang);
+    return fundsHelper.getFund(transaction.getFromFundId())
+      .thenCombine(fundsHelper.getFund(transaction.getToFundId()), (from, to) ->
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java:126]: It is possible to get both this funds in one query
`?query=id==transaction.getToFundId or id==transaction.getFromFundId`

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java (hunk 2 above)
Comment (siarhei-charniak) [reply src/main/java/org/folio/rest/helper/TransactionsHelper.java]: removed

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (siarhei-charniak) [reply descriptors/ModuleDescriptor-template.json]: removed

File: src/test/java/org/folio/rest/impl/TransactionTest.java (hunk 1 above)
Comment (siarhei-charniak) [reply src/test/java/org/folio/rest/impl/TransactionTest.java:52]: According to the [comments](https://issues.folio.org/browse/MODFIN-110?focusedCommentId=70531&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-70531), it should be allowed

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java (hunk 3 above)
Comment (siarhei-charniak) [reply src/main/java/org/folio/rest/helper/TransactionsHelper.java]: Same as [here](https://github.com/folio-org/mod-finance/pull/57#discussion_r420550425)

File: src/main/java/org/folio/rest/helper/TransactionsHelper.java [hunk 6]
```diff
@@ -90,4 +96,38 @@ private void validateTransactionType(Transaction transaction, Transaction.Transa
     return updateTransaction(transaction);
   }
 
+  private CompletableFuture<Transaction> checkRestrictions(Transaction transaction) {
+    CompletableFuture<Transaction> future = new VertxCompletableFuture<>(ctx);
+    switch(transaction.getTransactionType()) {
+      case ALLOCATION:
+      case TRANSFER:
+        if ((Objects.isNull(transaction.getFromFundId()) ^ Objects.isNull(transaction.getToFundId())) &&
```
Comment (AndreiMPV) [inline src/main/java/org/folio/rest/helper/TransactionsHelper.java:104]: Can we move this block of code to the method?

**Pull request merged**